### PR TITLE
Refactor Service Declaration Mgt for temporary SalesHeader/PurchHeader

### LINKED ISSUE #25222: [Bug]: codeunit 5012 "Service Declaration Mgt." - Hard Get On Header leads to problems with Temporary Record
### Describe the issue

Currently there are a few Event Subscribers in the `codeunit 5012 "Service Declaration Mgt."`, that make a hard get on the SalesHeader, PurchHeader instead of using the GetHeader Functions.

This leads to problems with temporary SalesHeader for price calculations

### Expected behavior

Those event subscribers should use the GetHeader methods

### Steps to reproduce

Create a temporary Record for a SalesHeader an and then create a tempory sales line for a resource.

### Additional context

Temporary SalesHeader and Purch Header are often forgotten.

### PR DESCRIPTION
**Pull request by pri-kise**
Fixes #25222

### CODE REVIEW

**Comment by pri-rapu** [comment]
LGTM - resolves errors for price calculation with temp. sales headers.

**Review by NKarolak — APPROVED** [review]

**Review by dannoe — APPROVED** [review]

**Comment by BertDeTemmerman** [comment]
Can this be checked and resolved please?
This is a blocking issue for us, and the only valid solution we found is deinstalling the app.
So we would like to have this in the next CU.

**Comment by JesperSchulz** [comment]
I will see what we can do! Processing PR!

File: Apps/W1/ServiceDeclaration/app/src/Processing/ServiceDeclarationMgt.Codeunit.al [hunk 1]
```diff
@@ -322,7 +322,7 @@ codeunit 5012 "Service Declaration Mgt."
         if not IsFeatureEnabled() then
             exit;
         if Item.Type = Item.Type::Service then begin
```
Comment (JesperSchulz) [inline Apps/W1/ServiceDeclaration/app/src/Processing/ServiceDeclarationMgt.Codeunit.al:324]: Why are we not just exiting here, if the SalesHeader is temporary? Would that not do the trick? Calling GetSalesHeader comes at a cost (does more than getting and invokes events).

**Review by JesperSchulz — COMMENTED** [review]
Just making sure we're doing the right thing. If you can confirm that "if IsTemporary() then exit" doesn't suffice, I'll rush this in - hopefully in time for 23.2 (today is last day for regular check-ins).

**Comment by duichwer** [comment]
Someone reported an issue with Standard Sales Code.
When they are set to automatic, then the lines are inserted before the SalesHeader is inserted.

Take a look on the SalesHeader Insert Trigger.

Therefore the GetSakesHeader is needed at least there.

File: Apps/W1/ServiceDeclaration/app/src/Processing/ServiceDeclarationMgt.Codeunit.al (hunk 1 above)
Comment (duichwer) [reply Apps/W1/ServiceDeclaration/app/src/Processing/ServiceDeclarationMgt.Codeunit.al:324]: See my comment below.

This PR does fix issues with Standard Sales Lines, too.

Additional we maybe can improve this in the future if we get a new method to simple calculate the Sales or Purch Prices for a Customer) Vendor.
Then we maybe won't need the TemSalesHeader anymore.

**Comment by JesperSchulz** [comment]
@duichwer, how about we do both. We exit when IsTemporary, but we use the GetSalesHeader instead of get. Any objections?

**Comment by JesperSchulz** [comment]
I have pushed the changes we are going to ship with 23.2. I hope those changes mostly do the trick. If not, could you please reopen the issue and create a new follow-up PR? I just really wanted to deliver something for 23.2, but the original PR got pushback, because it was calling events and functions not needed in this context. I'm sorry for overruling, but I thought it's better to deliver a solution to some of the problems, than none at all for 23.2. I hope that was ok.

**Comment by JesperSchulz** [comment]
Availability update: We will publish a fix for this issue in the next update for release 23.

Build ID to track: 14141.

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

Build ID: 14143.

**Review by AndreasMoth — APPROVED** [review]

**Pull request merged**